src/platform/datapath_win.c

```diff
@@ -0,0 +1,636 @@
+            QuicTraceLogVerbose(
+                RawSockCreateFail,
+                "[sock] Failed to create raw socket, status:%d", Status);
+            if (Datapath->UseTcp) {
+                CxPlatSocketDelete(*NewSocket);
+                goto Error;
+            }
+            Status = QUIC_STATUS_SUCCESS;
+        }
+    }
+
+Error:
+    return Status;
+}
+
+_IRQL_requires_max_(PASSIVE_LEVEL)
+QUIC_STATUS
+CxPlatSocketCreateTcp(
+    _In_ CXPLAT_DATAPATH* Datapath,
+    _In_opt_ const QUIC_ADDR* LocalAddress,
+    _In_ const QUIC_ADDR* RemoteAddress,
+    _In_opt_ void* CallbackContext,
+    _Out_ CXPLAT_SOCKET** NewSocket
+    )
+{
+    return SocketCreateTcp(
+        Datapath,
+        LocalAddress,
+        RemoteAddress,
+        CallbackContext,
+        NewSocket);
+}
+
+_IRQL_requires_max_(PASSIVE_LEVEL)
+QUIC_STATUS
+CxPlatSocketCreateTcpListener(
+    _In_ CXPLAT_DATAPATH* Datapath,
+    _In_opt_ const QUIC_ADDR* LocalAddress,
+    _In_opt_ void* RecvCallbackContext,
+    _Out_ CXPLAT_SOCKET** NewSocket
+    )
+{
+    return SocketCreateTcpListener(
+        Datapath,
+        LocalAddress,
+        RecvCallbackContext,
+        NewSocket);
+}
+
+_IRQL_requires_max_(PASSIVE_LEVEL)
+void
+CxPlatSocketDelete(
+    _In_ CXPLAT_SOCKET* Socket
+    )
+{
+    if (Socket->RawSocketAvailable) {
+        RawSocketDelete(SocketToRaw(Socket));
+    }
+    SocketDelete(Socket);
+}
+
+_IRQL_requires_max_(PASSIVE_LEVEL)
+QUIC_STATUS
+CxPlatSocketUpdateQeo(
+    _In_ CXPLAT_SOCKET* Socket,
+    _In_reads_(OffloadCount)
+        const CXPLAT_QEO_CONNECTION* Offloads,
+    _In_ uint32_t OffloadCount
+    )
+{
+    if (Socket->UseTcp || (Socket->RawSocketAvailable &&
+        !IS_LOOPBACK(Offloads[0].Address))) {
+        return RawSocketUpdateQeo(SocketToRaw(Socket), Offloads, OffloadCount);
+    }
+    return QUIC_STATUS_NOT_SUPPORTED;
+}
+
+_IRQL_requires_max_(DISPATCH_LEVEL)
+UINT16
+CxPlatSocketGetLocalMtu(
+    _In_ CXPLAT_SOCKET* Socket
+    )
+{
+    CXPLAT_DBG_ASSERT(Socket != NULL);
+    if (Socket->UseTcp || (Socket->RawSocketAvailable &&
+        !IS_LOOPBACK(Socket->RemoteAddress))) {
+        return RawSocketGetLocalMtu(SocketToRaw(Socket));
+    }
+    return Socket->Mtu;
+}
+
+_IRQL_requires_max_(DISPATCH_LEVEL)
+void
+CxPlatSocketGetLocalAddress(
+    _In_ CXPLAT_SOCKET* Socket,
+    _Out_ QUIC_ADDR* Address
+    )
+{
+    CXPLAT_DBG_ASSERT(Socket != NULL);
+    *Address = Socket->LocalAddress;
+}
+
+_IRQL_requires_max_(DISPATCH_LEVEL)
+void
+CxPlatSocketGetRemoteAddress(
+    _In_ CXPLAT_SOCKET* Socket,
+    _Out_ QUIC_ADDR* Address
+    )
+{
+    CXPLAT_DBG_ASSERT(Socket != NULL);
+    *Address = Socket->RemoteAddress;
+}
+
+_IRQL_requires_max_(DISPATCH_LEVEL)
+void
+CxPlatRecvDataReturn(
+    _In_opt_ CXPLAT_RECV_DATA* RecvDataChain
+    )
+{
+    if (RecvDataChain == NULL) {
+        return;
+    }
+
+    if (RecvDataChain->DatapathType == CXPLAT_DATAPATH_TYPE_USER) {
+        RecvDataReturn(RecvDataChain);
```

Comment:
Since you always check for NULL above, make sure these two functions don't also check, and update the SAL annotations to indicate no longer `_opt_`.